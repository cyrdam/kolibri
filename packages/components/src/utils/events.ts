// TODO: Should be synchronized with enums/events.ts
export type KoliBriEventType =
	| 'blur'
	| 'change'
	| 'changePage'
	| 'changePageSize'
	| 'click'
	| 'close'
	| 'focus'
	| 'input'
	| 'mousedown'
	| 'reset'
	| 'selection-change' // todo remove
	| 'selectionChange'
	| 'sort'
	| 'submit'
	| 'toggle';

export function stopPropagation(event: Event): void {
	event.stopImmediatePropagation();
	event.stopPropagation();
}

function createKoliBriEvent<T>(type: KoliBriEventType, detail: T): CustomEvent {
	const event = new CustomEvent(`kol-${type}`, {
		bubbles: true,
		cancelable: true,
		composed: true,
		detail: detail,
	});
	return event;
}

function dispatchKoliBriEvent<T>(target: EventTarget, type: KoliBriEventType, detail?: T): boolean {
	const dispatch = target.dispatchEvent(createKoliBriEvent(type, detail));
	return dispatch;
}

/**
 * @deprecated
 */
export function tryToDispatchKoliBriEvent<T>(type: KoliBriEventType, target?: EventTarget, detail?: T): void {
	target && dispatchKoliBriEvent(target, type, detail);
}

export function dispatchDomEvent(target: HTMLElement, type: KoliBriEventType) {
	target.dispatchEvent(new Event(type, { bubbles: true, composed: true }));
}

export function preventDefaultAndStopPropagation(event: Event) {
	event.preventDefault();
	stopPropagation(event);
}
