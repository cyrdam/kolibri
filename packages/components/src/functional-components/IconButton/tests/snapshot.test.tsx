import { h } from '@stencil/core';
import { renderFunctionalComponentToSpecPage } from '../../../utils/testing';
import type { IconButtonProps } from '../IconButton';
import KolIconButtonFc from '../IconButton';

describe('KolIconButtonFc', () => {
	it('should render button component correctly', async () => {
		const props: IconButtonProps = { componentName: 'button', label: 'Test Button', icon: 'test-icon' };
		const page = await renderFunctionalComponentToSpecPage(() => <KolIconButtonFc {...props} />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.tagName).toBe('KOL-BUTTON-WC');
		expect(page.root?.getAttribute('_label')).toContain('Test Button');
	});

	it('should render icon component correctly', async () => {
		const props: IconButtonProps = { componentName: 'icon', icon: 'test-icon' };
		const page = await renderFunctionalComponentToSpecPage(() => <KolIconButtonFc {...props} />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.tagName).toBe('KOL-ICON');
	});

	it('should handle onClick event', async () => {
		const onClick = jest.fn();
		const props: IconButtonProps = { componentName: 'button', label: 'Test Button', icon: 'test-icon', onClick };
		const page = await renderFunctionalComponentToSpecPage(() => <KolIconButtonFc {...props} />);
		expect(page.root?.getAttribute('onClick')).toBeDefined();
	});

	it('should render with additional props', async () => {
		const props: IconButtonProps = { componentName: 'button', label: 'Test Button', icon: 'test-icon', class: 'custom-class' };
		const page = await renderFunctionalComponentToSpecPage(() => <KolIconButtonFc {...props} />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.className).toContain('custom-class');
	});
});
