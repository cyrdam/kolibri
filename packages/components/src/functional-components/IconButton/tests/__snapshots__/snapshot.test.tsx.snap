// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KolIconButtonFc should render button component correctly 1`] = `
<kol-button-wc
  _hidelabel=""
  _icons="codicon codicon-test-icon"
  _label="Test Button"
  tabindex="0"
/>
`;

exports[`KolIconButtonFc should render icon component correctly 1`] = `
<kol-icon
  _icons="codicon codicon-test-icon"
  _label=""
/>
`;

exports[`KolIconButtonFc should render with additional props 1`] = `
<kol-button-wc
  _hidelabel=""
  _icons="codicon codicon-test-icon"
  _label="Test Button"
  class="custom-class"
  tabindex="0"
/>
`;
